File: backend/service/payment/src/main/java/com/leopoldhsing/digitalhippo/payment/api/ProductApi.java
package com.leopoldhsing.digitalhippo.payment.api;

import com.leopoldhsing.digitalhippo.model.dto.ErrorResponseDto;
import com.leopoldhsing.digitalhippo.model.entity.Product;
import com.leopoldhsing.digitalhippo.payment.service.ProductStripeService;
import com.stripe.exception.StripeException;
import io.swagger.v3.oas.annotations.*;
import io.swagger.v3.oas.annotations.media.*;
import io.swagger.v3.oas.annotations.responses.*;
import io.swagger.v3.oas.annotations.tags.Tag;
import org.springframework.web.bind.annotation.*;

@Tag(name = "Internal APIs for Stripe Product Management", description = "APIs for internal microservice communication. External access is restricted.")
@RestController
@RequestMapping("/api/stripe/inner/product")
public class ProductApi {

    private final ProductStripeService productStripeService;

    public ProductApi(ProductStripeService productStripeService) {
        this.productStripeService = productStripeService;
    }

    @Operation(
            summary = "Create a product in Stripe",
            description = "Creates a new product in Stripe based on the provided product information."
    )
    @ApiResponses(value = {
            @ApiResponse(
                    responseCode = "200",
                    description = "Product created successfully",
                    content = @Content(mediaType = "application/json", schema = @Schema(implementation = Product.class))
            ),
            @ApiResponse(
                    responseCode = "500",
                    description = "Stripe exception occurred",
                    content = @Content(schema = @Schema(implementation = ErrorResponseDto.class))
            )
    })
    @PostMapping
    public Product createStripeProduct(
            @Parameter(description = "Product information to create in Stripe", required = true)
            @RequestBody Product product
    ) {
        try {
            return productStripeService.createStripeProduct(product);
        } catch (StripeException e) {
            throw new RuntimeException(e);
        }
    }

    @Operation(
            summary = "Update a product in Stripe",
            description = "Updates an existing product in Stripe with the provided product information."
    )
    @ApiResponses(value = {
            @ApiResponse(
                    responseCode = "200",
                    description = "Product updated successfully",
                    content = @Content(mediaType = "application/json", schema = @Schema(implementation = Product.class))
            ),
            @ApiResponse(
                    responseCode = "500",
                    description = "Stripe exception occurred",
                    content = @Content(schema = @Schema(implementation = ErrorResponseDto.class))
            )
    })
    @PutMapping
    public Product updateStripeProduct(
            @Parameter(description = "Product information to update in Stripe", required = true)
            @RequestBody Product product
    ) {
        try {
            return productStripeService.updateStripeProduct(product);
        } catch (StripeException e) {
            throw new RuntimeException(e);
        }
    }

    @Operation(
            summary = "Delete a product from Stripe",
            description = "Deletes a product from Stripe using the provided Stripe product ID."
    )
    @ApiResponses(value = {
            @ApiResponse(
                    responseCode = "200",
                    description = "Product deleted successfully",
                    content = @Content(mediaType = "application/json", schema = @Schema(implementation = Boolean.class))
            ),
            @ApiResponse(
                    responseCode = "500",
                    description = "Stripe exception occurred",
                    content = @Content(schema = @Schema(implementation = ErrorResponseDto.class))
            )
    })
    @DeleteMapping("/{stripeId}")
    public Boolean deleteStripeProduct(
            @Parameter(description = "Stripe product ID to delete", required = true, example = "prod_ABC123")
            @PathVariable String stripeId
    ) {
        try {
            return productStripeService.deleteStripeProduct(stripeId);
        } catch (StripeException e) {
            throw new RuntimeException(e);
        }
    }
}

File: service/product/src/main/kotlin/com/leopoldhsing/digitalhippo/product/api/ProductApi.kt
package com.leopoldhsing.digitalhippo.product.api

import com.leopoldhsing.digitalhippo.model.dto.ErrorResponseDto
import com.leopoldhsing.digitalhippo.product.service.ProductService
import io.swagger.v3.oas.annotations.*
import io.swagger.v3.oas.annotations.media.*
import io.swagger.v3.oas.annotations.responses.*
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.*

@Tag(
    name = "Internal Product APIs",
    description = "APIs for internal microservice communication. External access is restricted."
)
@RestController
@RequestMapping("/api/product/inner")
open class ProductApi @Autowired constructor(private val productService: ProductService) {

    @Operation(
        summary = "Retrieve Stripe IDs for a list of product IDs",
        description = "Given a list of product IDs, return the corresponding Stripe product IDs."
    )
    @ApiResponses(
        value = [
            ApiResponse(
                responseCode = "200",
                description = "Stripe IDs retrieved successfully",
                content = [Content(mediaType = "application/json", array = ArraySchema(schema = Schema(type = "string")))]
            ),
            ApiResponse(
                responseCode = "404",
                description = "One or more products not found",
                content = [Content(schema = Schema(implementation = ErrorResponseDto::class))]
            )
        ]
    )
    @PostMapping("/stripe-ids")
    open fun getStripeIdList(
        @Parameter(description = "List of product IDs", required = true)
        @RequestBody productIdList: List<Long>
    ): List<String> {
        val products = productService.getProducts(productIdList)
        val stripeIdList: List<String> = products.map { it.stripeId }
        return stripeIdList
    }
}
